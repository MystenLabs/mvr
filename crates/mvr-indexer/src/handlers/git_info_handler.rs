use crate::handlers::MoveObjectProcessor;
use crate::models::mainnet::sui::dynamic_field::Field;
use crate::models::{mainnet, testnet};
use anyhow::Error;
use async_trait::async_trait;
use diesel::query_dsl::methods::FilterDsl;
use diesel::upsert::excluded;
use diesel::ExpressionMethods;
use diesel_async::RunQueryDsl;
use move_types::MoveStruct;
use mvr_schema::models::GitInfo;
use serde::de::DeserializeOwned;
use std::marker::PhantomData;
use std::sync::Arc;
use sui_indexer_alt_framework::pipeline::concurrent::Handler;
use sui_indexer_alt_framework::pipeline::Processor;
use sui_pg_db::Connection;
use sui_types::base_types::MoveObjectType;
use sui_types::full_checkpoint_content::CheckpointData;
use crate::handlers::convert_struct_tag;

pub type MainnetGitInfo = Field<u64, mainnet::mvr_metadata::git::GitInfo>;
pub type TestnetGitInfo = Field<u64, testnet::mvr_metadata::git::GitInfo>;

pub struct GitInfoHandler<T> {
    chain_id: String,
    type_: MoveObjectType,
    phantom_data: PhantomData<T>,
}

impl<T: MoveStruct> GitInfoHandler<T> {
    pub fn new(chain_id: String) -> Self {
        // Indexing dynamic field object Field<u64, [metadata_pkg_id]::git::GitInfo>
        let struct_tag = Field::<u64, MoveGitInfo>::struct_type();
        GitInfoHandler {
            chain_id,
            type_: MoveObjectType::from(convert_struct_tag(struct_tag)),
            phantom_data: Default::default(),
        }
    }
}

impl MoveObjectProcessor<MainnetGitInfo, GitInfo> for GitInfoHandler<MainnetGitInfo> {
    const PROC_NAME: &'static str = "GitInfo - Mainnet";

    fn process_move_object(
        chain_id: String,
        move_obj: MainnetGitInfo,
        obj: &Object,
    ) -> Result<GitInfo, Error> {
        let mainnet::mvr_metadata::git::GitInfo {
            repository,
            path,
            tag,
        } = move_obj.value;
        Ok(GitInfo {
            table_id: obj.owner.get_owner_address()?.to_string(),
            object_version: obj.version().value() as i64,
            version: move_obj.name as i32,
            chain_id,
            repository: Some(repository),
            path: Some(path),
            tag: Some(tag),
        })
    }
}

impl MoveObjectProcessor<TestnetGitInfo, GitInfo> for GitInfoHandler<TestnetGitInfo> {
    const PROC_NAME: &'static str = "GitInfo - Testnet";

    fn process_move_object(
        chain_id: String,
        move_obj: TestnetGitInfo,
        obj: &Object,
    ) -> Result<GitInfo, Error> {
        let testnet::mvr_metadata::git::GitInfo {
            repository,
            path,
            tag,
        } = move_obj.value;
        Ok(GitInfo {
            table_id: obj.owner.get_owner_address()?.to_string(),
            object_version: obj.version().value() as i64,
            version: move_obj.name as i32,
            chain_id,
            repository: Some(repository),
            path: Some(path),
            tag: Some(tag),
        })
    }
}

#[async_trait]
impl<T: MoveStruct + DeserializeOwned> Handler for GitInfoHandler<T>
where
    Self: MoveObjectProcessor<T, GitInfo>,
{
    async fn commit(values: &[Self::Value], conn: &mut Connection<'_>) -> anyhow::Result<usize> {
        use mvr_schema::schema::git_infos::columns::*;
        use mvr_schema::schema::git_infos::dsl::git_infos;
        Ok(diesel::insert_into(git_infos)
            .values(values)
            .on_conflict((table_id, version))
            .do_update()
            .set((
                object_version.eq(excluded(object_version)),
                repository.eq(excluded(repository)),
                path.eq(excluded(path)),
                tag.eq(excluded(tag)),
            ))
            .filter(object_version.lt(excluded(object_version)))
            .execute(conn)
            .await?)
    }
}

impl<T: MoveStruct + DeserializeOwned> Processor for GitInfoHandler<T>
where
    Self: MoveObjectProcessor<T, GitInfo>,
{
    const NAME: &'static str = Self::PROC_NAME;
    type Value = GitInfo;

    fn process(&self, checkpoint: &Arc<CheckpointData>) -> anyhow::Result<Vec<Self::Value>> {
        checkpoint
            .transactions
            .iter()
            .try_fold(vec![], |result, tx| {
                tx.output_objects.iter().try_fold(result, |mut result, o| {
                    if matches!(o.type_(), Some(t) if t == &self.type_) {
                        if let Some(move_obj) = o.data.try_as_move() {
                            result.push(Self::process_move_object(
                                self.chain_id.clone(),
                                bcs::from_bytes(move_obj.contents())?,
                                o,
                            )?)
                        }
                    }
                    Ok(result)
                })
            })
    }
}
