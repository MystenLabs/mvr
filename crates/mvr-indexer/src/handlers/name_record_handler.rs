use crate::models::mvr_core;
use crate::models::mvr_core::name::Name;
use async_trait::async_trait;
use diesel::upsert::excluded;
use diesel::ExpressionMethods;
use diesel_async::RunQueryDsl;
use itertools::Itertools;
use mvr_schema::models::NameRecord;
use std::sync::Arc;
use sui_indexer_alt_framework::pipeline::concurrent::Handler;
use sui_indexer_alt_framework::pipeline::Processor;
use sui_pg_db::Connection;
use sui_types::base_types::MoveObjectType;
use sui_types::dynamic_field::{DynamicFieldInfo, Field};
use sui_types::full_checkpoint_content::CheckpointData;

const TESTNET_KEY: &str = "testnet";

pub struct NameRecordHandler {
    type_: MoveObjectType,
}

impl NameRecordHandler {
    pub fn new() -> Self {
        // Indexing dynamic field object Field<[mvr_core]::name::Name, [mvr_core]::app_record::AppRecord>
        let name_type = mvr_core::name::Name::type_();
        let app_record = mvr_core::app_record::AppRecord::type_();
        let type_ = MoveObjectType::from(DynamicFieldInfo::dynamic_field_type(
            name_type.into(),
            app_record.into(),
        ));
        NameRecordHandler { type_ }
    }
}
#[async_trait]
impl Handler for NameRecordHandler {
    async fn commit(values: &[Self::Value], conn: &mut Connection<'_>) -> anyhow::Result<usize> {
        use mvr_schema::schema::name_records::dsl::name_records;
        use mvr_schema::schema::name_records::columns::*;
        Ok(diesel::insert_into(name_records)
            .values(values)
            .on_conflict(name)
            .do_update()
            .set((
                mainnet_id.eq(excluded(mainnet_id)),
                testnet_id.eq(excluded(testnet_id)),
                metadata.eq(excluded(metadata))
            ))
            .execute(conn)
            .await?)
    }
}

impl Processor for NameRecordHandler {
    const NAME: &'static str = "NameRecord";
    type Value = NameRecord;

    fn process(&self, checkpoint: &Arc<CheckpointData>) -> anyhow::Result<Vec<Self::Value>> {
        checkpoint
            .transactions
            .iter()
            .try_fold(vec![], |result, tx| {
                tx.output_objects.iter().try_fold(result, |mut result, o| {
                    if matches!(o.type_(), Some(t) if t == &self.type_) {
                        if let Some(move_obj) = o.data.try_as_move() {
                            use crate::models::mvr_core::app_record::AppRecord as MoveAppRecord;
                            let data: Field<Name, MoveAppRecord> =
                                bcs::from_bytes(move_obj.contents())?;
                            let Name { org, app } = data.name;
                            let MoveAppRecord {
                                app_info,
                                networks,
                                metadata,
                                ..
                            } = data.value;
                            result.push(NameRecord {
                                // TODO: use correct name to string
                                name: format!("{org:?}/{}", app.join("/")),
                                mainnet_id: app_info
                                    .and_then(|info| Some(info.package_info_id?.to_string())),
                                testnet_id: networks
                                    .get(&TESTNET_KEY.into())
                                    .and_then(|info| Some(info.package_info_id?.to_string())),
                                metadata: serde_json::to_value(metadata)?,
                            })
                        }
                    }
                    Ok(result)
                })
            })
    }
}
