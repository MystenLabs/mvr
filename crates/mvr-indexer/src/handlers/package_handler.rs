use crate::{MAINNET_CHAIN_ID, TESTNET_CHAIN_ID};
use chrono::DateTime;
use diesel_async::RunQueryDsl;
use itertools::Itertools;
use move_binary_format::CompiledModule;
use mvr_schema::models::{Package, PackageDependency};
use std::collections::HashSet;
use std::sync::Arc;
use sui_indexer_alt_framework::pipeline::concurrent::Handler;
use sui_indexer_alt_framework::pipeline::Processor;
use sui_pg_db::{Connection, Db};
use sui_types::base_types::ObjectID;
use sui_types::full_checkpoint_content::CheckpointData;
use sui_types::move_package::MovePackage;
use sui_types::object::Data;

pub struct PackageHandler<const MAINNET: bool>;

impl<const MAINNET: bool> PackageHandler<MAINNET> {
    const CHAIN_ID: &'static str = if MAINNET {
        MAINNET_CHAIN_ID
    } else {
        TESTNET_CHAIN_ID
    };
}

#[async_trait::async_trait]
impl<const MAINNET: bool> Handler for PackageHandler<MAINNET> {
    type Store = Db;

    async fn commit<'a>(
        values: &[Self::Value],
        conn: &mut Connection<'a>,
    ) -> anyhow::Result<usize> {
        use mvr_schema::schema::package_dependencies::dsl::package_dependencies;
        use mvr_schema::schema::packages::dsl::packages;

        let deps = values.iter().flat_map(|p| &p.deps).collect::<Vec<_>>();

        let insert_package = diesel::insert_into(packages)
            .values(values)
            .on_conflict_do_nothing()
            .execute(conn)
            .await?;

        let insert_deps = diesel::insert_into(package_dependencies)
            .values(deps)
            .on_conflict_do_nothing()
            .execute(conn)
            .await?;

        Ok(insert_package + insert_deps)
    }
}

impl<const MAINNET: bool> Processor for PackageHandler<MAINNET> {
    const NAME: &'static str = if MAINNET {
        "Package - Mainnet"
    } else {
        "Package - Testnet"
    };
    type Value = Package;

    fn process(&self, checkpoint: &Arc<CheckpointData>) -> anyhow::Result<Vec<Self::Value>> {
        let timestamp =
            DateTime::from_timestamp_millis(checkpoint.checkpoint_summary.timestamp_ms as i64)
                .unwrap()
                .naive_utc();

        let mut results = vec![];
        for tx in &checkpoint.transactions {
            for o in &tx.output_objects {
                if let Data::Package(p) = &o.data {
                    let package_id = p.id().to_hex_uncompressed();
                    let deps = package_dependencies(Self::CHAIN_ID.to_string(), p)?;
                    results.push(Package {
                        package_id,
                        original_id: p.original_package_id().to_hex_uncompressed(),
                        package_version: p.version().value() as i64,
                        move_package: bcs::to_bytes(p)?,
                        chain_id: Self::CHAIN_ID.to_string(),
                        tx_hash: tx.transaction.digest().base58_encode(),
                        sender: tx.transaction.sender_address().to_string(),
                        timestamp,
                        deps,
                    })
                }
            }
        }
        Ok(results)
    }
}

pub fn package_dependencies(
    chain_id: String,
    package: &MovePackage,
) -> Result<Vec<PackageDependency>, anyhow::Error> {
    let package_id = package.id().to_hex_uncompressed();
    let mut immediate_dependencies = HashSet::new();
    for (_, bytes) in package.serialized_module_map() {
        let module = CompiledModule::deserialize_with_defaults(bytes)?;
        immediate_dependencies.extend(
            module
                .immediate_dependencies()
                .into_iter()
                .map(|dep| ObjectID::from(*dep.address())),
        );
    }
    Ok(package
        .linkage_table()
        .iter()
        .map(|(_, u)| u.upgraded_id)
        .dedup()
        .map(|dependency_package_id| PackageDependency {
            package_id: package_id.clone(),
            dependency_package_id: dependency_package_id.to_hex_uncompressed(),
            chain_id: chain_id.clone(),
            immediate_dependency: immediate_dependencies.contains(&dependency_package_id),
        })
        .collect())
}
