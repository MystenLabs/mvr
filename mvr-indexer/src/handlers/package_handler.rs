use crate::models::{Package, SuiEnv};
use diesel_async::RunQueryDsl;
use std::sync::Arc;
use sui_indexer_alt_framework::pipeline::concurrent::Handler;
use sui_indexer_alt_framework::pipeline::Processor;
use sui_pg_db::Connection;
use sui_sdk::types::full_checkpoint_content::{CheckpointData, CheckpointTransaction};

pub struct PackageHandler {
    sui_env: SuiEnv,
}

impl PackageHandler {
    pub fn new(sui_env: SuiEnv) -> Self {
        Self { sui_env }
    }
}

#[async_trait::async_trait]
impl Handler for PackageHandler {
    async fn commit(values: &[Self::Value], conn: &mut Connection<'_>) -> anyhow::Result<usize> {
        use crate::schema::packages::dsl::packages;
        Ok(diesel::insert_into(packages)
            .values(values)
            .on_conflict_do_nothing()
            .execute(conn)
            .await?)
    }
}

impl Processor for PackageHandler {
    const NAME: &'static str = "Package";
    type Value = Package;

    fn process(&self, checkpoint: &Arc<CheckpointData>) -> anyhow::Result<Vec<Self::Value>> {
        // todo: store sui env as well in case we need to clean up testnet data after wipe.
        checkpoint
            .transactions
            .iter()
            .try_fold(vec![], |mut results, tx| {
                if contain_package(tx) {
                    results.append(
                        &mut tx
                            .output_objects
                            .iter()
                            .map(|o| {
                                o.data
                                    .try_as_package()
                                    .iter()
                                    .map(|p| {
                                        Ok::<_, anyhow::Error>(Package {
                                            package_id: p.id().to_hex_literal(),
                                            original_id: p.original_package_id().to_hex_literal(),
                                            package_version: p.version().value() as i64,
                                            move_package: bcs::to_bytes(p)?,
                                        })
                                    })
                                    .collect::<Vec<_>>()
                            })
                            .flatten()
                            .collect::<anyhow::Result<Vec<_>>>()?,
                    )
                }
                Ok(results)
            })
    }
}

fn contain_package(tx: &CheckpointTransaction) -> bool {
    tx.output_objects.iter().any(|o| o.is_package())
}
