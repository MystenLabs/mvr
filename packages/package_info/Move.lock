# @generated by Move, please check-in and do not edit manually.

[move]
version = 2
manifest_digest = "6EAF9E5D09667BDD9425A1DA5C58682515E7136A6DC12E3F5013480775A38409"
deps_digest = "F8BBB0CCB2491CA29A3DF03D6F92277A4F3574266507ACD77214D37ECA3F3082"
dependencies = [
  { name = "Sui" },
]

[[move.package]]
name = "MoveStdlib"
source = { git = "https://github.com/MystenLabs/sui.git", rev = "framework/testnet", subdir = "crates/sui-framework/packages/move-stdlib" }

[[move.package]]
name = "Sui"
source = { git = "https://github.com/MystenLabs/sui.git", rev = "framework/testnet", subdir = "crates/sui-framework/packages/sui-framework" }

dependencies = [
  { name = "MoveStdlib" },
]

[move.toolchain-version]
compiler-version = "1.24.0"
edition = "2024.beta"
flavor = "sui"

[env]

[env.testnet]
chain-id = "4c78adac"
original-published-id = "0x08ad4adcb87d72b0c1626ee1c50cd1c0a9642c703b91df49399f44880d774841"
latest-published-id = "0x08ad4adcb87d72b0c1626ee1c50cd1c0a9642c703b91df49399f44880d774841"
published-version = "1"

[env.mainnet]
chain-id = "35834a8a"
original-published-id = "0x15f29668bc8a8168975f6de8da926ea398619cfff5894bb71d31a777fa7ea18e"
latest-published-id = "0x15f29668bc8a8168975f6de8da926ea398619cfff5894bb71d31a777fa7ea18e"
published-version = "1"
