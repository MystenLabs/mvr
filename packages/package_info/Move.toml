[package]
name = "package_info"
edition = "2024.beta"

[dependencies]
Sui = { git = "https://github.com/MystenLabs/sui.git", subdir = "crates/sui-framework/packages/sui-framework", rev = "framework/testnet", override=true }

# goes into [dependencies] section
Codec = { git = "https://github.com/sui-potatoes/app.git", subdir = "packages/codec", rev = "codec@mainnet-v1" }


# goes into [dependencies] section
# Codec = { git = "https://github.com/sui-potatoes/app.git", subdir = "packages/codec", rev = "codec@testnet-v1" }

[addresses]
package_info="0x0"
