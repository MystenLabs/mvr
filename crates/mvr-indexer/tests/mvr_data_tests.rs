use chrono::NaiveDateTime;
use fastcrypto::hash::{HashFunction, Sha256};
use insta::assert_json_snapshot;
use mvr_indexer::handlers::git_info_handler::GitInfoHandler;
use mvr_indexer::handlers::git_info_handler::MainnetGitInfo;
use mvr_indexer::handlers::name_record_handler::NameRecordHandler;
use mvr_indexer::handlers::package_handler::PackageHandler;
use mvr_indexer::handlers::package_info_handler::PackageInfoHandler;
use mvr_indexer::models::mainnet::mvr_metadata::package_info::PackageInfo;
use mvr_indexer::MAINNET_CHAIN_ID;
use mvr_schema::MIGRATIONS;
use serde_json::Value;
use sqlx::{Column, PgPool, Row, ValueRef};
use std::fs;
use std::path::Path;
use std::sync::Arc;
use sui_indexer_alt_framework::pipeline::concurrent::Handler;
use sui_indexer_alt_framework::pipeline::Processor;
use sui_indexer_alt_framework::store::Store;
use sui_pg_db::temp::TempDb;
use sui_pg_db::Connection;
use sui_pg_db::Db;
use sui_pg_db::DbArgs;
use sui_storage::blob::Blob;
use sui_types::full_checkpoint_content::CheckpointData;

#[tokio::test]
async fn git_info_write_test() -> Result<(), anyhow::Error> {
    let handler = GitInfoHandler::<MainnetGitInfo>::new(MAINNET_CHAIN_ID.into());
    data_test("git_infos_write", handler, ["git_infos"]).await?;
    Ok(())
}

#[tokio::test]
async fn package_write_test() -> Result<(), anyhow::Error> {
    let handler = PackageHandler::<true>;
    data_test(
        "packages_write",
        handler,
        ["packages", "package_dependencies"],
    )
    .await?;
    Ok(())
}

#[tokio::test]
async fn package_infos_write_test() -> Result<(), anyhow::Error> {
    let handler = PackageInfoHandler::<PackageInfo>::new(MAINNET_CHAIN_ID.into());
    data_test("package_infos_write", handler, ["package_infos"]).await?;
    Ok(())
}

#[tokio::test]
async fn name_records_write_test() -> Result<(), anyhow::Error> {
    let handler = NameRecordHandler::new();
    data_test("name_records_write", handler, ["name_records"]).await?;
    Ok(())
}

#[tokio::test]
async fn immediate_dependency_test() -> Result<(), anyhow::Error> {
    let handler = PackageHandler::<true>;
    data_test(
        "immediate_dependency",
        handler,
        ["packages", "package_dependencies"],
    )
    .await?;
    Ok(())
}

async fn data_test<H, I>(
    test_name: &str,
    handler: H,
    tables_to_check: I,
) -> Result<(), anyhow::Error>
where
    I: IntoIterator<Item = &'static str>,
    H: Handler + Processor,
    for<'a> H::Store: Store<Connection<'a> = Connection<'a>>,
{
    // Set up the temporary database
    let temp_db = TempDb::new()?;
    let url = temp_db.database().url();
    let db = Arc::new(Db::for_write(url.clone(), DbArgs::default()).await?);
    db.run_migrations(MIGRATIONS).await?;
    let mut conn = db.connect().await?;

    // Test setup based on provided test_name
    let test_path = Path::new("tests/checkpoints").join(test_name);
    let checkpoints = get_checkpoints_in_folder(&test_path)?;

    // Run pipeline for each checkpoint
    for checkpoint in checkpoints {
        run_pipeline(&handler, &checkpoint, &mut conn).await?;
    }

    // Check results by comparing database tables with snapshots
    for table in tables_to_check {
        let rows = read_table(&table, &url.to_string()).await?;
        assert_json_snapshot!(format!("{test_name}__{table}"), rows);
    }
    Ok(())
}

async fn run_pipeline<'c, T: Handler + Processor, P: AsRef<Path>>(
    handler: &T,
    path: P,
    conn: &mut Connection<'c>,
) -> Result<(), anyhow::Error>
where
    T::Store: Store<Connection<'c> = Connection<'c>>,
{
    let bytes = fs::read(path)?;
    let cp = Blob::from_bytes::<CheckpointData>(&bytes)?;
    let result = handler.process(&Arc::new(cp))?;
    T::commit(&result, conn).await?;
    Ok(())
}

/// Read the entire table from database as json value.
/// note: bytea values will be hashed to reduce output size.
async fn read_table(table_name: &str, db_url: &str) -> Result<Vec<Value>, anyhow::Error> {
    let pool = PgPool::connect(db_url).await?;
    let rows = sqlx::query(&format!("SELECT * FROM {table_name}"))
        .fetch_all(&pool)
        .await?;

    // To json
    Ok(rows
        .iter()
        .map(|row| {
            let mut obj = serde_json::Map::new();

            for column in row.columns() {
                let column_name = column.name();

                let value = if let Ok(v) = row.try_get::<String, _>(column_name) {
                    Value::String(v)
                } else if let Ok(v) = row.try_get::<i32, _>(column_name) {
                    Value::Number(v.into())
                } else if let Ok(v) = row.try_get::<i64, _>(column_name) {
                    Value::Number(v.into())
                } else if let Ok(v) = row.try_get::<bool, _>(column_name) {
                    Value::Bool(v)
                } else if let Ok(v) = row.try_get::<Value, _>(column_name) {
                    v
                } else if let Ok(v) = row.try_get::<Vec<u8>, _>(column_name) {
                    // hash bytea contents
                    let mut hash_function = Sha256::default();
                    hash_function.update(v);
                    let digest2 = hash_function.finalize();
                    Value::String(digest2.to_string())
                } else if let Ok(v) = row.try_get::<NaiveDateTime, _>(column_name) {
                    Value::String(v.to_string())
                } else if let Ok(true) = row.try_get_raw(column_name).map(|v| v.is_null()) {
                    Value::Null
                } else {
                    panic!(
                        "Cannot parse DB value to json, type: {:?}, column: {column_name}",
                        row.try_get_raw(column_name)
                            .map(|v| v.type_info().to_string())
                    )
                };
                obj.insert(column_name.to_string(), value);
            }

            Value::Object(obj)
        })
        .collect())
}

fn get_checkpoints_in_folder(folder: &Path) -> Result<Vec<String>, anyhow::Error> {
    let mut files = Vec::new();

    // Read the directory
    for entry in fs::read_dir(folder)? {
        let entry = entry?;
        let path = entry.path();

        // Check if it's a file and ends with ".chk"
        if path.is_file() && path.extension().and_then(|s| s.to_str()) == Some("chk") {
            files.push(path.display().to_string());
        }
    }

    Ok(files)
}
