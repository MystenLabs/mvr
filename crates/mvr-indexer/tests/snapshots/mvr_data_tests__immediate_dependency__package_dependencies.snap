---
source: crates/mvr-indexer/tests/mvr_data_tests.rs
expression: rows
---
[
  {
    "package_id": "0x0ce90717ec5f301c84c0be83893ab8cb64e6fe574f627c040b16ec4b1f87fbb6",
    "dependency_package_id": "0x0000000000000000000000000000000000000000000000000000000000000001",
    "chain_id": "35834a8a",
    "immediate_dependency": true
  },
  {
    "package_id": "0x0ce90717ec5f301c84c0be83893ab8cb64e6fe574f627c040b16ec4b1f87fbb6",
    "dependency_package_id": "0x0000000000000000000000000000000000000000000000000000000000000002",
    "chain_id": "35834a8a",
    "immediate_dependency": true
  },
  {
    "package_id": "0x0ce90717ec5f301c84c0be83893ab8cb64e6fe574f627c040b16ec4b1f87fbb6",
    "dependency_package_id": "0x0000000000000000000000000000000000000000000000000000000000000003",
    "chain_id": "35834a8a",
    "immediate_dependency": true
  },
  {
    "package_id": "0x0ce90717ec5f301c84c0be83893ab8cb64e6fe574f627c040b16ec4b1f87fbb6",
    "dependency_package_id": "0x000000000000000000000000000000000000000000000000000000000000dee9",
    "chain_id": "35834a8a",
    "immediate_dependency": true
  },
  {
    "package_id": "0x0ce90717ec5f301c84c0be83893ab8cb64e6fe574f627c040b16ec4b1f87fbb6",
    "dependency_package_id": "0x0b572349baf4526c92c4e5242306e07a1658fa329ae93d1b9db0fc38b8a592bb",
    "chain_id": "35834a8a",
    "immediate_dependency": true
  },
  {
    "package_id": "0x0ce90717ec5f301c84c0be83893ab8cb64e6fe574f627c040b16ec4b1f87fbb6",
    "dependency_package_id": "0xdc67d6de3f00051c505da10d8f6fbab3b3ec21ec65f0dc22a2f36c13fc102110",
    "chain_id": "35834a8a",
    "immediate_dependency": false
  },
  {
    "package_id": "0x0ce90717ec5f301c84c0be83893ab8cb64e6fe574f627c040b16ec4b1f87fbb6",
    "dependency_package_id": "0x25929e7f29e0a30eb4e692952ba1b5b65a3a4d65ab5f2a32e1ba3edcb587f26d",
    "chain_id": "35834a8a",
    "immediate_dependency": true
  },
  {
    "package_id": "0x0ce90717ec5f301c84c0be83893ab8cb64e6fe574f627c040b16ec4b1f87fbb6",
    "dependency_package_id": "0x2c8d603bc51326b8c13cef9dd07031a408a48dddb541963357661df5d3204809",
    "chain_id": "35834a8a",
    "immediate_dependency": true
  },
  {
    "package_id": "0x0ce90717ec5f301c84c0be83893ab8cb64e6fe574f627c040b16ec4b1f87fbb6",
    "dependency_package_id": "0x2d9316f1f1a95f6d7c85a4e690ef7c359e6649773ef2c37ad7d9857adb6bef06",
    "chain_id": "35834a8a",
    "immediate_dependency": true
  },
  {
    "package_id": "0x0ce90717ec5f301c84c0be83893ab8cb64e6fe574f627c040b16ec4b1f87fbb6",
    "dependency_package_id": "0x549e8b69270defbfafd4f94e17ec44cdbdd99820b33bda2278dea3b9a32d3f55",
    "chain_id": "35834a8a",
    "immediate_dependency": true
  },
  {
    "package_id": "0x0ce90717ec5f301c84c0be83893ab8cb64e6fe574f627c040b16ec4b1f87fbb6",
    "dependency_package_id": "0x64213b0e4a52bac468d4ac3f140242f70714381653a1919a6d57cd49c628207a",
    "chain_id": "35834a8a",
    "immediate_dependency": true
  },
  {
    "package_id": "0x0ce90717ec5f301c84c0be83893ab8cb64e6fe574f627c040b16ec4b1f87fbb6",
    "dependency_package_id": "0x712579292f80c11a0c9de4ff553d6e5c4757105e83a8a3129823d2b39e65d062",
    "chain_id": "35834a8a",
    "immediate_dependency": false
  },
  {
    "package_id": "0x0ce90717ec5f301c84c0be83893ab8cb64e6fe574f627c040b16ec4b1f87fbb6",
    "dependency_package_id": "0xe2b515f0052c0b3f83c23db045d49dbe1732818ccfc5d4596c9482f7f2e76a85",
    "chain_id": "35834a8a",
    "immediate_dependency": false
  },
  {
    "package_id": "0x0ce90717ec5f301c84c0be83893ab8cb64e6fe574f627c040b16ec4b1f87fbb6",
    "dependency_package_id": "0x73baa782c55003b3a359dec04b189312565d18e7309d4a51f5f112f891e3b2ab",
    "chain_id": "35834a8a",
    "immediate_dependency": false
  },
  {
    "package_id": "0x0ce90717ec5f301c84c0be83893ab8cb64e6fe574f627c040b16ec4b1f87fbb6",
    "dependency_package_id": "0x7f6ce7ade63857c4fd16ef7783fed2dfc4d7fb7e40615abdb653030b76aef0c6",
    "chain_id": "35834a8a",
    "immediate_dependency": true
  },
  {
    "package_id": "0x0ce90717ec5f301c84c0be83893ab8cb64e6fe574f627c040b16ec4b1f87fbb6",
    "dependency_package_id": "0x8e6adb40c50cf2ce81d7dd12d240fb881df715343ab61233124aadb87a9fd1fd",
    "chain_id": "35834a8a",
    "immediate_dependency": false
  },
  {
    "package_id": "0x0ce90717ec5f301c84c0be83893ab8cb64e6fe574f627c040b16ec4b1f87fbb6",
    "dependency_package_id": "0x1a3c42ded7b75cdf4ebc7c7b7da9d1e1db49f16fcdca934fac003f35f39ecad9",
    "chain_id": "35834a8a",
    "immediate_dependency": false
  },
  {
    "package_id": "0x0ce90717ec5f301c84c0be83893ab8cb64e6fe574f627c040b16ec4b1f87fbb6",
    "dependency_package_id": "0xa0eba10b173538c8fecca1dff298e488402cc9ff374f8a12ca7758eebe830b66",
    "chain_id": "35834a8a",
    "immediate_dependency": true
  },
  {
    "package_id": "0x0ce90717ec5f301c84c0be83893ab8cb64e6fe574f627c040b16ec4b1f87fbb6",
    "dependency_package_id": "0xa6baab1e668c7868991c1c3c11e144100f5734c407d020f72a01b9d1a8bcb97f",
    "chain_id": "35834a8a",
    "immediate_dependency": true
  },
  {
    "package_id": "0x0ce90717ec5f301c84c0be83893ab8cb64e6fe574f627c040b16ec4b1f87fbb6",
    "dependency_package_id": "0xb196672db3293fdebdbd4cbea950823ff84805547c7345710a1cf9d0db52938f",
    "chain_id": "35834a8a",
    "immediate_dependency": true
  },
  {
    "package_id": "0x0ce90717ec5f301c84c0be83893ab8cb64e6fe574f627c040b16ec4b1f87fbb6",
    "dependency_package_id": "0x7be61b62d902f3fe78d0a5e20b81b4715a47ff06cae292db8991dfea422cf57e",
    "chain_id": "35834a8a",
    "immediate_dependency": false
  },
  {
    "package_id": "0x0ce90717ec5f301c84c0be83893ab8cb64e6fe574f627c040b16ec4b1f87fbb6",
    "dependency_package_id": "0xba153169476e8c3114962261d1edc70de5ad9781b83cc617ecc8c1923191cae0",
    "chain_id": "35834a8a",
    "immediate_dependency": true
  },
  {
    "package_id": "0x0ce90717ec5f301c84c0be83893ab8cb64e6fe574f627c040b16ec4b1f87fbb6",
    "dependency_package_id": "0x1d56b8ec33c3fae897eb7bb1acb79914e8152faed614868928e684c25c8b198d",
    "chain_id": "35834a8a",
    "immediate_dependency": false
  },
  {
    "package_id": "0x0ce90717ec5f301c84c0be83893ab8cb64e6fe574f627c040b16ec4b1f87fbb6",
    "dependency_package_id": "0xe93247b408fe44ed0ee5b6ac508b36325b239d6333e44ffa240dcc0c1a69cdd8",
    "chain_id": "35834a8a",
    "immediate_dependency": false
  },
  {
    "package_id": "0x0ce90717ec5f301c84c0be83893ab8cb64e6fe574f627c040b16ec4b1f87fbb6",
    "dependency_package_id": "0xc66fabf1a9253e43c70f1cc02d40a1d18db183140ecaae2a3f58fa6b66c55acf",
    "chain_id": "35834a8a",
    "immediate_dependency": true
  },
  {
    "package_id": "0x0ce90717ec5f301c84c0be83893ab8cb64e6fe574f627c040b16ec4b1f87fbb6",
    "dependency_package_id": "0xceb3b6f35b71dbd0296cd96f8c00959c230854c7797294148b413094b9621b0e",
    "chain_id": "35834a8a",
    "immediate_dependency": false
  },
  {
    "package_id": "0x0ce90717ec5f301c84c0be83893ab8cb64e6fe574f627c040b16ec4b1f87fbb6",
    "dependency_package_id": "0xdb982f402a039f196f3e13cd73795db441393b5bc6eef7a0295a333808982a7d",
    "chain_id": "35834a8a",
    "immediate_dependency": false
  },
  {
    "package_id": "0x0ce90717ec5f301c84c0be83893ab8cb64e6fe574f627c040b16ec4b1f87fbb6",
    "dependency_package_id": "0xdeeb7a4662eec9f2f3def03fb937a663dddaa2e215b8078a284d026b7946c270",
    "chain_id": "35834a8a",
    "immediate_dependency": true
  },
  {
    "package_id": "0x0ce90717ec5f301c84c0be83893ab8cb64e6fe574f627c040b16ec4b1f87fbb6",
    "dependency_package_id": "0xc4049b2d1cc0f6e017fda8260e4377cecd236bd7f56a54fee120816e72e2e0dd",
    "chain_id": "35834a8a",
    "immediate_dependency": false
  },
  {
    "package_id": "0x0ce90717ec5f301c84c0be83893ab8cb64e6fe574f627c040b16ec4b1f87fbb6",
    "dependency_package_id": "0xf325ce1300e8dac124071d3152c5c5ee6174914f8bc2161e88329cf579246efc",
    "chain_id": "35834a8a",
    "immediate_dependency": true
  },
  {
    "package_id": "0x0ce90717ec5f301c84c0be83893ab8cb64e6fe574f627c040b16ec4b1f87fbb6",
    "dependency_package_id": "0xbd8d4489782042c6fafad4de4bc6a5e0b84a43c6c00647ffd7062d1e2bb7549e",
    "chain_id": "35834a8a",
    "immediate_dependency": false
  }
]
